File: apps/ios/FarCooler/VTCore.swift
import Foundation
import FarCoolerVT

/// Swift's view of the Rust terminal core.
///
/// A thin, safe wrapper: it owns the handle's lifetime, converts between Swift
/// and C types, and nothing else. Every decision about what bytes mean — what
/// color an escape sequence produces, what an arrow key encodes to under the
/// program's current mode — lives on the other side of this boundary. That is
/// what let the Mac app's renderer and this one answer to the same handle
/// without agreeing on a single line of emulator logic between them.
///
/// Not thread-safe by design — the core is confined to the main actor, which
/// is where drawing happens anyway, and is the same discipline the header
/// documents for every platform.
@MainActor
final class VTCore {
    /// The pointer is just an address — it carries no isolated state of its
    /// own. The discipline that matters is that only the main actor ever calls
    /// through it, and deinit is by definition the last thing that does.
    nonisolated(unsafe) private var handle: UnsafeMutableRawPointer?

    init(columns: Int, rows: Int) {
        handle = farcooler_vt_new(UInt16(clamping: columns), UInt16(clamping: rows))
    }

    deinit {
        if let handle { farcooler_vt_free(handle) }
    }

    func feed(_ bytes: [UInt8]) {
        guard let handle, !bytes.isEmpty else { return }
        bytes.withUnsafeBufferPointer { farcooler_vt_feed(handle, $0.baseAddress, $0.count) }
    }

    /// Recolour every cell the next snapshot produces.
    ///
    /// Nineteen packed values: sixteen ANSI, then foreground, background,
    /// cursor. Colours resolve when a snapshot is taken rather than when bytes
    /// arrive, so this recolours scrollback too.
    @discardableResult
    func setPalette(_ colors: [UInt32]) -> Bool {
        guard let handle else { return false }
        return colors.withUnsafeBufferPointer {
            farcooler_vt_set_palette(handle, $0.baseAddress, $0.count)
        }
    }

    func resize(columns: Int, rows: Int) {
        guard let handle else { return }
        farcooler_vt_resize(handle, UInt16(clamping: columns), UInt16(clamping: rows))
    }

    /// Scroll the view. Positive goes back into history.
    func scroll(lines: Int32) {
        guard let handle else { return }
        farcooler_vt_scroll(handle, lines)
    }

    /// Jump back to the live screen. Call this on input: typing into a
    /// scrolled-back view would show the user nothing of what they typed.
    func scrollToBottom() {
        guard let handle else { return }
        farcooler_vt_scroll_to_bottom(handle)
    }

    /// Read the screen and hand it to `body`.
    ///
    /// Scoped rather than returned: the cell buffer belongs to the core and is
    /// only valid until the next call on this handle. Handing it out directly
    /// would invite a dangling read the moment a poll fed new bytes; a closure
    /// makes the lifetime the compiler's problem instead of a bug report.
    func withSnapshot<T>(_ body: (VTSnapshot) -> T) -> T? {
        guard let handle else { return nil }
        var raw = FarCoolerVtSnapshot()
        guard farcooler_vt_snapshot(handle, &raw), let cells = raw.cells else { return nil }
        let count = Int(raw.rows) * Int(raw.columns)
        return body(
            VTSnapshot(
                cells: UnsafeBufferPointer(start: cells, count: count),
                columns: Int(raw.columns),
                rows: Int(raw.rows),
                cursorRow: Int(raw.cursor_row),
                cursorColumn: Int(raw.cursor_column),
                cursorVisible: raw.cursor_visible,
                displayOffset: Int(raw.display_offset),
                historySize: Int(raw.history_size)
            )
        )
    }

    /// How far back the view is scrolled, and how much there is to scroll
    /// through. Both zero on a core that has only ever been fed a captured
    /// screen — which is the whole shape of the scrolling bug, and why this is
    /// worth a call of its own rather than being read off a grid.
    ///
    /// `withSnapshot` already carries both, but every caller of it is building
    /// a `TerminalGrid` to draw. Asking "where are we" should not require
    /// copying four thousand cells to find out.
    var scrollPosition: (offset: Int, history: Int) {
        withSnapshot { ($0.displayOffset, $0.historySize) } ?? (0, 0)
    }

    /// Whether the program has swapped to the alternate screen.
    ///
    /// The alternate screen has no scrollback by definition — it is a fresh
    /// buffer the program paints and later discards, which is why leaving
    /// `vim` puts your shell back exactly as it was. So it is the one place
    /// where "scroll" cannot mean "look at what came before": there is nothing
    /// before it, and the wheel can only sensibly go to the program.
    ///
    /// Read live rather than cached: a program enters and leaves it whenever
    /// it likes, and a stale answer sends a swipe to the wrong place.
    var isAlternateScreen: Bool {
        guard let handle else { return false }
        return farcooler_vt_alt_screen(handle)
    }

    /// Whether the program running in this pane has asked for mouse events.
    ///
    /// Asked by trying to encode one and seeing whether the core produces
    /// bytes: `farcooler_vt_encode_mouse` returns nothing when the program has
    /// mouse reporting off, which makes "would this encode" and "does the
    /// program want the mouse" the same question. No new FFI for a fact the
    /// core already answers.
    ///
    /// It exists to be OBSERVED, not to decide anything. `TerminalSession.scroll`
    /// deliberately no longer branches on it — that was the bug — and it is
    /// published on the surface so a UI test can tell the two panes apart. A
    /// pane with mouse reporting on and one without look identical on a screen,
    /// and the difference between them was the whole of "scroll is broken for
    /// terminals": untestable while nothing could say which kind of pane was in
    /// front of you.
    var wantsMouse: Bool {
        encode(
            mouse: UInt32(FARCOOLER_VT_MOUSE_WHEEL_UP),
            action: UInt32(FARCOOLER_VT_MOUSE_PRESS),
            column: 0, row: 0, modifiers: []) != nil
    }

    /// Text the program asked to put on the clipboard (OSC 52), or nil.
    ///
    /// This matters more here than it does on the Mac: there is no text
    /// selection in this renderer, so OSC 52 is the only way for anything on
    /// screen to reach the clipboard at all.
    ///
    /// Two calls, because the core reports the size it needs and writes nothing
    /// when the buffer is short rather than truncating — half a copied command
    /// is worse than no copy. There is no counterpart that READS the clipboard;
    /// a program asking for its contents is refused inside the core.
    func takeClipboard() -> String? {
        guard let handle else { return nil }
        let needed = farcooler_vt_take_clipboard(handle, nil, 0)
        guard needed > 0 else { return nil }

        var buffer = [UInt8](repeating: 0, count: needed)
        let written = buffer.withUnsafeMutableBufferPointer {
            farcooler_vt_take_clipboard(handle, $0.baseAddress, $0.count)
        }
        guard written == needed else { return nil }
        return String(decoding: buffer, as: UTF8.self)
    }

    /// The URL under a cell, or nil.
    ///
    /// The core decides what counts as a URL and which schemes may be opened.
    /// Terminal output is not trusted input — an agent prints whatever it read —
    /// and keeping the allowlist there means it is one list rather than three.
    func url(atRow row: Int, column: Int) -> String? {
        guard let handle, row >= 0, column >= 0 else { return nil }
        var span = FarCoolerVtUrlSpan()
        let needed = farcooler_vt_url_at(
            handle, UInt16(clamping: row), UInt16(clamping: column), &span, nil, 0)
        guard needed > 0 else { return nil }

        var buffer = [UInt8](repeating: 0, count: needed)
        let written = buffer.withUnsafeMutableBufferPointer {
            farcooler_vt_url_at(
                handle, UInt16(clamping: row), UInt16(clamping: column), &span,
                $0.baseAddress, $0.count)
        }
        guard written == needed else { return nil }
        return String(decoding: buffer, as: UTF8.self)
    }

    /// Encode a keystroke for the program currently running.
    ///
    /// The core answers because the answer depends on modes it holds: an arrow
    /// key is different bytes under application cursor mode, and a Ctrl
    /// modifier turns a printable scalar into a control code (Ctrl-C is `c`
    /// with `.control`, not a separate key). Neither is this file's to decide.
    func encode(key: UInt32, modifiers: VTModifiers) -> [UInt8] {
        guard let handle else { return [] }
        var buffer = [UInt8](repeating: 0, count: 16)
        let n = buffer.withUnsafeMutableBufferPointer {
            farcooler_vt_encode_key(handle, key, modifiers.rawValue, $0.baseAddress, $0.count)
        }
        return Array(buffer[0..<n])
    }

    func encode(scalar: Unicode.Scalar, modifiers: VTModifiers) -> [UInt8] {
        encode(key: scalar.value, modifiers: modifiers)
    }

    /// Encode a mouse event, or nil when the program does not want it — in
    /// which case the event is ours to handle locally.
    func encode(
        mouse button: UInt32,
        action: UInt32,
        column: Int,
        row: Int,
        modifiers: VTModifiers
    ) -> [UInt8]? {
        guard let handle else { return nil }
        var buffer = [UInt8](repeating: 0, count: 32)
        let n = buffer.withUnsafeMutableBufferPointer {
            farcooler_vt_encode_mouse(
                handle,
                button,
                action,
                UInt16(clamping: column),
                UInt16(clamping: row),
                modifiers.rawValue,
                $0.baseAddress,
                $0.count
            )
        }
        return n > 0 ? Array(buffer[0..<n]) : nil
    }
}

/// A borrowed view of the screen. Valid only inside `withSnapshot`.
struct VTSnapshot {
    let cells: UnsafeBufferPointer<FarCoolerVtCell>
    let columns: Int
    let rows: Int
    /// Where the emulator itself has the caret, after every byte it has been
    /// given.
    ///
    /// Worth taking over the host's answer to `terminal.cursor` whenever there
    /// is one: that is a second round trip whose answer is true at a different
    /// instant than the screen it is drawn on, so a fast-moving caret is drawn
    /// a poll behind the text it is supposed to be sitting in. This one cannot
    /// disagree with the screen beside it, because the same bytes put both
    /// where they are.
    let cursorRow: Int
    let cursorColumn: Int
    let cursorVisible: Bool
    /// How many lines above the live screen the view is showing. Zero is live.
    let displayOffset: Int
    /// How many lines sit above the screen and could be scrolled to.
    ///
    /// Zero means a swipe has nowhere to go, and saying so is the difference
    /// between a scroll that does nothing and a scroll that cannot do
    /// anything. A capture-only core reports zero here forever; see
    /// `TerminalSession.render`.
    let historySize: Int

    subscript(row: Int, column: Int) -> FarCoolerVtCell {
        cells[row * columns + column]
    }
}

struct VTModifiers: OptionSet {
    let rawValue: UInt32
    static let shift = VTModifiers(rawValue: UInt32(FARCOOLER_VT_MOD_SHIFT))
    static let alt = VTModifiers(rawValue: UInt32(FARCOOLER_VT_MOD_ALT))
    static let control = VTModifiers(rawValue: UInt32(FARCOOLER_VT_MOD_CTRL))
}

extension FarCoolerVtCell {
    var isBold: Bool { flags & UInt16(FARCOOLER_VT_FLAG_BOLD) != 0 }
    var isInverse: Bool { flags & UInt16(FARCOOLER_VT_FLAG_INVERSE) != 0 }
    /// A double-width character. The next column is its spacer; a renderer
    /// that draws both would show the glyph twice.
    var isWide: Bool { flags & UInt16(FARCOOLER_VT_FLAG_WIDE) != 0 }

    /// Nil for a blank cell, so a renderer can skip drawing rather than
    /// measuring and painting an empty glyph four thousand times a frame.
    var character: Character? {
        guard let scalar = Unicode.Scalar(ch), scalar != " " else { return nil }
        return Character(scalar)
    }
}

File: apps/android/app/src/main/java/com/farcooler/core/VtCore.kt
package com.farcooler.core

/**
 * Kotlin's view of the Rust terminal core.
 *
 * A thin, safe wrapper: it owns the handle's lifetime, converts between Kotlin
 * and the flat arrays the JNI layer hands back, and nothing else. Every
 * decision about what bytes mean — what colour an escape sequence produces,
 * what an arrow key encodes to under the program's current mode — lives on the
 * other side of this boundary. That is what lets this renderer and the Mac's
 * answer to the same core without agreeing on a single line of emulator logic.
 *
 * Not thread-safe by design, matching the contract the C header documents for
 * every renderer. [com.farcooler.net.TerminalSession] confines it to one
 * dispatcher.
 */
class VtCore(columns: Int, rows: Int) {
    private var handle: Long =
        if (NativeLibrary.loaded) NativeVt.nativeNew(columns, rows) else 0

    fun free() {
        if (handle == 0L) return
        NativeVt.nativeFree(handle)
        handle = 0
    }

    fun feed(bytes: ByteArray) {
        if (handle == 0L || bytes.isEmpty()) return
        NativeVt.nativeFeed(handle, bytes)
    }

    fun resize(columns: Int, rows: Int) {
        if (handle == 0L) return
        NativeVt.nativeResize(handle, columns, rows)
    }

    /** Scroll the view. Positive goes back into history. */
    fun scroll(lines: Int) {
        if (handle == 0L) return
        NativeVt.nativeScroll(handle, lines)
    }

    /**
     * Jump back to the live screen. Call this on input: typing into a
     * scrolled-back view would show the user nothing of what they typed.
     */
    /**
     * Recolour every cell the next snapshot produces.
     *
     * Colours resolve when a snapshot is taken rather than when bytes arrive,
     * so this recolours scrollback too — switching themes is instant, not a
     * repaint of history the terminal no longer holds.
     */
    fun setPalette(colors: IntArray): Boolean {
        val h = handle
        if (h == 0L) return false
        return NativeVt.nativeSetPalette(h, colors)
    }

    fun scrollToBottom() {
        if (handle == 0L) return
        NativeVt.nativeScrollToBottom(handle)
    }

    /** A copy of the screen, safe to hold and hand to Compose. */
    fun snapshot(): TerminalGrid? {
        if (handle == 0L) return null
        val flat = NativeVt.nativeSnapshot(handle) ?: return null
        return TerminalGrid.from(flat)
    }

    /**
     * Bytes the program wants written back to the pty.
     *
     * Cursor-position reports and mouse replies must reach the program or a
     * full-screen agent sits waiting for an answer that never comes.
     */
    fun takeWrites(): ByteArray {
        if (handle == 0L) return ByteArray(0)
        return NativeVt.nativeTakeWrites(handle) ?: ByteArray(0)
    }

    fun title(): String? = if (handle == 0L) null else NativeVt.nativeTitle(handle)

    /**
     * Text the program asked to put on the clipboard (OSC 52), or null.
     *
     * This matters more here than on the Mac: there is no text selection in
     * this renderer, so OSC 52 is the only way anything on screen reaches the
     * clipboard at all.
     *
     * There is no counterpart that reads the clipboard. A program asking for
     * its contents is refused inside the core — copy is a program handing you
     * something, paste is a program taking something, and Far Cooler runs
     * agents on machines nobody is watching.
     */
    fun takeClipboard(): String? =
        if (handle == 0L) null else NativeVt.nativeTakeClipboard(handle)

    /**
     * The URL under a cell of the screen as currently shown, or null.
     *
     * The core decides what counts as a URL and which schemes may be opened;
     * terminal output is not trusted input, and keeping the allowlist there
     * makes it one list rather than three.
     */
    fun urlAt(row: Int, column: Int): String? =
        if (handle == 0L || row < 0 || column < 0) null
        else NativeVt.nativeUrlAt(handle, row, column)

    /**
     * Encode a keystroke for the program currently running.
     *
     * The core answers because the answer depends on modes it holds: an arrow
     * key is different bytes under application cursor mode, and a Ctrl modifier
     * turns a printable scalar into a control code (Ctrl-C is `c` with
     * [Vt.MOD_CTRL], not a separate key). Neither is this file's to decide.
     */
    fun encodeKey(key: Int, modifiers: Int): ByteArray {
        if (handle == 0L) return ByteArray(0)
        return NativeVt.nativeEncodeKey(handle, key, modifiers) ?: ByteArray(0)
    }

    /** Null when the program does not want the event — handle it locally. */
    fun encodeMouse(button: Int, action: Int, column: Int, row: Int, modifiers: Int): ByteArray? {
        if (handle == 0L) return null
        return NativeVt.nativeEncodeMouse(handle, button, action, column, row, modifiers)
    }

    /**
     * Encode pasted text, bracketing it if the program asked for that —
     * without which an editor auto-indents every pasted line and a shell runs
     * each newline as a command.
     */
    fun encodePaste(text: String): ByteArray {
        if (handle == 0L) return ByteArray(0)
        return NativeVt.nativeEncodePaste(handle, text) ?: ByteArray(0)
    }
}

/**
 * A plain-data copy of one screen.
 *
 * The JNI layer hands back one flat `int[]` rather than an array of objects,
 * and it is unpacked here into four parallel arrays rather than a list of cell
 * objects, for the same reason: an 80×24 screen is 1,920 cells redrawn several
 * times a second, and per-cell objects are exactly the shape of garbage that
 * makes a scrolling terminal stutter.
 */
class TerminalGrid(
    val columns: Int,
    val rows: Int,
    val cursorRow: Int,
    val cursorColumn: Int,
    val cursorVisible: Boolean,
    val displayOffset: Int,
    val historySize: Int,
    private val characters: IntArray,
    private val foregrounds: IntArray,
    private val backgrounds: IntArray,
    private val flags: IntArray,
) {
    private fun index(row: Int, column: Int) = row * columns + column

    /** The Unicode scalar in a cell, or 0 where nothing was written. */
    fun character(row: Int, column: Int): Int = characters[index(row, column)]

    /**
     * Already resolved for [Vt.FLAG_INVERSE] — the core reports foreground and
     * background separately from the flag, and a renderer that forgot to swap
     * them would draw reverse-video text invisibly on itself.
     */
    fun foreground(row: Int, column: Int): Int {
        val i = index(row, column)
        val packed = if (inverse(i)) backgrounds[i] else foregrounds[i]
        return opaque(packed)
    }

    fun background(row: Int, column: Int): Int {
        val i = index(row, column)
        val packed = if (inverse(i)) foregrounds[i] else backgrounds[i]
        return opaque(packed)
    }

    fun bold(row: Int, column: Int): Boolean = flags[index(row, column)] and Vt.FLAG_BOLD != 0

    /**
     * A double-width character. The next column is its spacer; a renderer that
     * drew both would show the glyph twice.
     */
    fun wide(row: Int, column: Int): Boolean = flags[index(row, column)] and Vt.FLAG_WIDE != 0

    /**
     * The same screen with the caret somewhere else.
     *
     * For a screen built from a capture rather than a live stream: a capture is
     * text, so feeding it leaves the emulator's caret wherever the last
     * character landed — the bottom left — rather than at the prompt someone is
     * typing into. The host reports where it really is, and this is where that
     * answer is applied. The cell arrays are shared rather than copied; nothing
     * mutates them after construction.
     */
    fun withCursor(row: Int, column: Int) = TerminalGrid(
        columns = columns,
        rows = rows,
        cursorRow = row.coerceIn(0, (rows - 1).coerceAtLeast(0)),
        cursorColumn = column.coerceIn(0, (columns - 1).coerceAtLeast(0)),
        cursorVisible = cursorVisible,
        displayOffset = displayOffset,
        historySize = historySize,
        characters = characters,
        foregrounds = foregrounds,
        backgrounds = backgrounds,
        flags = flags,
    )

    private fun inverse(i: Int) = flags[i] and Vt.FLAG_INVERSE != 0

    private fun opaque(packed: Int) = packed or 0xFF000000.toInt()

    companion object {
        private const val HEADER = 7

        fun from(flat: IntArray): TerminalGrid? {
            if (flat.size < HEADER) return null
            val columns = flat[0]
            val rows = flat[1]
            val count = columns * rows
            if (columns <= 0 || rows <= 0 || flat.size < HEADER + count * 4) return null

            val characters = IntArray(count)
            val foregrounds = IntArray(count)
            val backgrounds = IntArray(count)
            val flags = IntArray(count)
            var read = HEADER
            for (i in 0 until count) {
                characters[i] = flat[read++]
                foregrounds[i] = flat[read++]
                backgrounds[i] = flat[read++]
                flags[i] = flat[read++]
            }

            return TerminalGrid(
                columns = columns,
                rows = rows,
                // Clamped rather than trusted: the cursor position can come
                // from a separate call than the screen dump (see
                // `TerminalSession.render`), and the two can race a resize by
                // one poll interval.
                cursorRow = flat[2].coerceIn(0, (rows - 1).coerceAtLeast(0)),
                cursorColumn = flat[3].coerceIn(0, (columns - 1).coerceAtLeast(0)),
                cursorVisible = flat[4] != 0,
                displayOffset = flat[5],
                historySize = flat[6],
                characters = characters,
                foregrounds = foregrounds,
                backgrounds = backgrounds,
                flags = flags,
            )
        }
    }
}

/**
 * Colours the core did not resolve, because they belong to this screen rather
 * than to any program running on the host: the fill behind a short last row and
 * the cursor block. Values mirror the Mac and iOS apps' so the same terminal
 * looks like the same terminal on all three.
 */
/**
 * The colours the renderer owns, read from the theme in force.
 *
 * These were constants, which is what made the palette unchangeable without
 * reinstalling. Cell colours still come from the core already resolved; this is
 * the chrome around them.
 */
object TerminalPalette {
    val BACKGROUND: Int get() =
        com.farcooler.data.Themes.opaque(com.farcooler.data.Themes.current.background)
    val CURSOR: Int get() =
        com.farcooler.data.Themes.opaque(com.farcooler.data.Themes.current.cursor)
}
